andela/mrm_api
# #163751049 Refactor backend codebase to align to new structure

#### What does this PR do?
Refactor codebase to align with structure documented [here](https://docs.google.com/document/d/1dowMWUNUxJvcFqBiR4y7TCCZYDSqNgfryK1k8wnvm0M/edit), that is, remove Office, Block, Floor and Wing entities.

#### Description of Task to be completed?
The current codebase contains Offices, Blocks, Floors and Wing entities. These have been removed and replaced with Tags.

#### How should this be manually tested?
* Clone the branch and follow the following steps to [setup](https://github.com/andela/mrm_api).
* git checkout to `ft-refactor-backend-structure-163751049` and run all mutations and queries in the codebase to the new version out.

* Due to the latest changes in the tables/models follow the steps below to run migrations:
```- Checkout to develop
- Clear alembic_version table.
- Run migrations `alembic stamp head && alembic revision --autogenerate -m "message && alembic upgrade head"` or `"make migrate message="message"`
- Delete generated file.
- Run `alembic stamp head` or add latest-revision number in alembic_versions table if using docker.
- Checkout to branch with changes.
- Run `alembic upgrade head` or `make migrate`
- Related tables should be removed on checking the db.
```
Mutation to create room as in the newly implemented codebase:
```
mutation {
        createRoom(
            name: "Hogwarts", roomType: "Meeting", capacity: 4, roomTags: [1], locationId:1, calendarId:"andela.com_3836323338323230343935@resource.calendar.google.com",
            imageUrl: "https://www.officelovin.com/wp-content/uploads/2016/10/andela-office-main-1.jpg") { 
            room {
                id
                name
              	locationId
                roomType
                capacity
                imageUrl
                roomTags {
                  name
                  description
                }
            }
        }
    }
```

#### Any background context you want to provide?
An admin can add/update one or more tags in a room. Deleting a room automatically deletes the relationship between the room and its tags.

#### Screenshots
* On creating a room:
<img width="1231" alt="screenshot 2019-02-08 at 09 50 10" src="https://user-images.githubusercontent.com/26184534/52468222-1e4f1280-2b99-11e9-802f-494100eeee7c.png">

* On getting all rooms
<img width="1440" alt="screenshot 2019-02-08 at 09 57 43" src="https://user-images.githubusercontent.com/26184534/52468229-25762080-2b99-11e9-8ce8-00b1a476b62d.png">

* On getting a room by its id
<img width="1440" alt="screenshot 2019-02-08 at 10 03 00" src="https://user-images.githubusercontent.com/26184534/52468237-28711100-2b99-11e9-927b-dea6c6cfc085.png">


#### What are the relevant pivotal tracker stories?
[#163751049](https://www.pivotaltracker.com/story/show/163751049)
